kernel/src/table_changes/scan.rs

```diff
@@ -0,0 +1,203 @@
+                if self
+                    .table_changes
+                    .partition_columns()
+                    .contains(logical_field.name())
+                {
+                    // Store the index into the schema for this field. When we turn it into an
+                    // expression in the inner loop, we will index into the schema and get the name and
+                    // data type, which we need to properly materialize the column.
+                    have_partition_cols = true;
+                    Ok(ColumnType::Partition(index))
+                } else if CDF_FIELDS
+                    .iter()
+                    .any(|field| field.name() == logical_field.name())
+                {
+                    // CDF Columns are generated, so they do not have a column mapping.
+                    Ok(ColumnType::Selected(logical_field.name().to_string()))
+                } else {
+                    // Add to read schema, store field so we can build a `Column` expression later
+                    // if needed (i.e. if we have partition columns)
+                    let physical_field =
+                        logical_field.make_physical(*self.table_changes.column_mapping_mode())?;
+                    debug!("\n\n{logical_field:#?}\nAfter mapping: {physical_field:#?}\n\n");
+                    let physical_name = physical_field.name.clone();
+                    read_fields.push(physical_field);
+                    Ok(ColumnType::Selected(physical_name))
+                }
+            })
+            .try_collect()?;
+        Ok(TableChangesScan {
+            table_changes: self.table_changes,
+            logical_schema,
+            predicate: self.predicate,
+            all_fields: column_types,
+            have_partition_cols,
+        })
+    }
+}
+#[cfg(test)]
+mod tests {
+
+    use std::sync::Arc;
+
+    use crate::engine::sync::SyncEngine;
+    use crate::expressions::{column_expr, Scalar};
+    use crate::scan::ColumnType;
+    use crate::schema::{DataType, StructField, StructType};
+    use crate::{Expression, Table};
+
+    #[test]
+    fn simple_table_changes_scan_builder() {
+        let path = "./tests/data/table-with-cdf";
+        let engine = Box::new(SyncEngine::new());
+        let table = Table::try_from_uri(path).unwrap();
+
+        // A field in the schema goes from being nullable to non-nullable
+        let table_changes = table.table_changes(engine.as_ref(), 0, 1).unwrap();
+
+        let scan = table_changes.into_scan_builder().build().unwrap();
+        assert_eq!(
+            scan.all_fields,
+            vec![
+                ColumnType::Selected("part".to_string()),
+                ColumnType::Selected("id".to_string()),
+                ColumnType::Selected("_change_type".to_string()),
+                ColumnType::Selected("_commit_version".to_string()),
+                ColumnType::Selected("_commit_timestamp".to_string()),
```

Comment:
I guess these generated columns are not file-constant values, so we have to treat them like normal columns even tho they don't come from the parquet?

Reply:
Correct. I considered introducing a `ColumnType::Generated`. I feel like this is the "right" way, but may take some more discussion and changes to `kernel/scan/mod.rs`. If you think `ColumnType::Generated` is not controversial, we can go with that.

The current plan is to treat them as selected, and later check when transforming physical to logical:
```
match (column_type) {
    ColumnType::Selected(col) => {
       if CDF_FIELDS.contains(col) {
           // Treat as CDF generated column
       } else {
           // Usual path for ColumnType::Selected
       }
   }
   ...
}
```

Reply:
@nicklan @zachschuermann  If you feel strongly that we should do `Generated` or keep `Selected`, let me know!

Reply:
Yeah, and I think the expression stuff will change this as well. Rather than a "selected" field, we'll just generate an expression to say, add a column with this value.

Reply:
sure SGTM :)